changelog_generator/repository_manager.py

```diff
@@ -1,101 +1,156 @@
 import re
-from typing import Iterable, List, Optional
+from abc import ABC, abstractmethod
+from functools import lru_cache
+from typing import ClassVar, Iterable, List, Pattern, Sequence, Tuple
 
 from git import Repo
 
 from changelog_generator.commit import Commit
 
-tag_version_re = re.compile(
-    r"^v?(?P<major>\d+)[-\.]+(?P<minor>\d+)[-\.]+((?P<bug>\d+)|rc(?P<rc>\d+)){0,1}$"
-)
 
-remote_re = re.compile(
-    r"^(https://github\.com/|git@github\.com:)"
-    r"(?P<organization>[^/]+)/(?P<repository>[^.]+)(\.git)?$"
-)
+class BaseTagManager(ABC):
+    PATTERN: ClassVar[Pattern[str]]
+
+    def get_semver_from_tag(self, tag: str) -> Tuple[int, ...]:
+        """
+        Splits a tag in a semver tuple of ints (major,minor,bug,rc).
+        """
+        res = self.PATTERN.match(tag)
+        if not res:
+            raise ValueError(f"Not a valid tag: {tag}")
+        return tuple(
+            map(
+                int,
+                [
+                    res.group("major"),
+                    res.group("minor"),
+                    res.group("bug") or 0,
+                    res.group("rc") or 0,
+                ],
+            )
+        )
 
+    @abstractmethod
+    def is_release_tag(self, tag: str) -> bool:
+        """validates a tag"""
+
+    @abstractmethod
+    def get_tags(self):
+        """list the tags"""
+
+    def get_release_tags(self) -> Sequence[str]:
+        """Lists only the release tags"""
+        return tuple(
+            sorted(
+                filter(self.is_release_tag, self.get_tags()),
+                key=self.get_semver_from_tag,
+                reverse=True,
+            )
+        )
 
-def is_tag_version(tag: str) -> bool:
-    res = tag_version_re.match(tag)
-    if not res:
-        return False
-    return not res.group("rc")
 
+class SimpleTagManager(BaseTagManager):
+    """
+    Matches tags of the kind:
+        - 1.2.3
+        - 1-2-3
+        - v1.2.3
+        - 1.2.3rc1
+    """
 
-def get_tag_value(tag: str) -> int:
-    res = tag_version_re.match(tag)
-    assert res
-    return -1 * (
-        1000000 * int(res.group("major"))
-        + 10000 * int(res.group("minor"))
-        + 100 * int(res.group("bug") or "0")
-        + int(res.group("rc") or "100")
+    PATTERN = re.compile(
+        r"^v?(?P<major>\d+)[-.](?P<minor>\d+)[-.]((?P<bug>\d+)|rc(?P<rc>\d+))?$"
     )
 
+    def __init__(self, repository: Repo) -> None:
+        self.repository = repository
+
+    def is_release_tag(self, tag: str) -> bool:
+        res = self.PATTERN.match(tag)
+        return res and not res.group("rc")
+
+    def get_tags(self):
+        return self.repository.git.tag(f"--merged").split("\n")
+
+
+class PrefixedTagManager(BaseTagManager):
+    """
+    Matches tags of the kind:
+        - toto/1.2.3
+        - toto/1.2.3rc1
+    """
+
+    PATTERN = re.compile(
+        r"^(?P<prefix>\w+)/(?P<major>\d+)\.(?P<minor>\d+)\.((?P<bug>\d+)|rc(?P<rc>\d+))?$"
```

Comment:
Why not simply using an optional group ?
```
r"^(?:(?P<prefix>\w+)/)?(?P<major>\d+)\.(?P<minor>\d+)\.((?P<bug>\d+)|rc(?P<rc>\d+))?$"
```

Reply:
I find it easier to maintain to have classes. If someone wants to add a new kind of tag parsing/filtering/sorting, he will just have to add a other class, and the rest should still work